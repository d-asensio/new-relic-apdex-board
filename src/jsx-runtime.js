export default function createElement (tag, props, ...children) {
  const element = document.createElement(tag)

  if (props !== null) {
    _addPropsToElement(props, element)
  }

  _appendChildrenToElement(children, element)

  return element
}

function _addPropsToElement (props, element) {
  const { className, ...attributes } = props

  _addAttributesToElement(attributes, element)

  if (className) {
    _addClassNameToElement(className, element)
  }
}

function _addAttributesToElement (attributes, element) {
  const attributeEntries = Object.entries(attributes)

  for (const [attrName, attrValue] of attributeEntries) {
    _attributeSupportedByElementOrThrow(attrName, element)

    element.setAttribute(attrName, attrValue)
  }
}

function _attributeSupportedByElementOrThrow (attributeName, element) {
  if (!_isAttributeSupportedByElement(attributeName, element)) {
    throw new Error(
      `A "${element.tagName.toLowerCase()}" element do not support the attribute "${attributeName}".`
    )
  }
}

function _isAttributeSupportedByElement (attributeName, element) {
  if (_isDataAttribute(attributeName)) return true
  if (_isAriaAttribute(attributeName)) return true

  return attributeName in element
}

function _isDataAttribute (attributeName) {
  return attributeName.startsWith('data-')
}

function _isAriaAttribute (attributeName) {
  return attributeName.startsWith('aria-')
}

function _addClassNameToElement (className, element) {
  const sanitizedClassNames = _sanitizeClassName(className)

  for (const sanitizedClass of sanitizedClassNames) {
    element.classList.add(sanitizedClass)
  }
}

function _sanitizeClassName (className) {
  return className
    .split(' ')
    .filter(Boolean)
}

function _appendChildrenToElement (children, element) {
  for (const child of children) {
    if (Array.isArray(child)) {
      _appendChildrenToElement(child, element)
    } else {
      _appendChildToElement(child, element)
    }
  }
}

function _appendChildToElement (child, element) {
  const safeChildNode = _createSafeChildNode(child, element)

  element.appendChild(safeChildNode)
}

function _createSafeChildNode (child) {
  if (child instanceof HTMLElement || child instanceof Text) {
    return child
  }

  return document.createTextNode(child)
}
