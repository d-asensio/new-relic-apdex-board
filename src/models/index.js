export * from './application.model.js'
export * from './dashboard.model.js'
export * from './host.model.js'
