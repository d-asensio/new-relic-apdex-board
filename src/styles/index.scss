@import "reset";

@import "tokens/all";
@import "utilities/all";

@import "views/all";

body {
  background: #f5f5f5;
  
  @include sp-inset($space--large);
}
