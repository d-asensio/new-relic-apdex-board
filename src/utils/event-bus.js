export class EventBus {
  constructor () {
    this._events = new Map()
  }

  on (eventName, listener) {
    this._isValidListenerOrThrow(listener)
    this._registerEventListener(eventName, listener)

    return () => {
      this.off(eventName, listener)
    }
  }

  off (eventName, listener) {
    this._isValidListenerOrThrow(listener)
    this._unregisterEventListener(eventName, listener)
  }

  async emit (eventName, ...eventData) {
    await this._callEventListeners(eventName, eventData)
  }

  _registerEventListener (eventName, listener) {
    const eventListeners = this._getOrCreateEventListeners(eventName)

    eventListeners.add(listener)
  }

  _unregisterEventListener (eventName, listener) {
    const eventListeners = this._getOrCreateEventListeners(eventName)

    eventListeners.delete(listener)
  }

  async _callEventListeners (eventName, eventData) {
    const eventListeners = this._getOrCreateEventListeners(eventName)

    for (const listener of eventListeners) {
      await (async function IIFE (closuredEventData) {
        listener(...closuredEventData)
      })(eventData)
    }
  }

  _isValidListenerOrThrow (listener) {
    if (typeof listener !== 'function') {
      throw new TypeError(
        'The event listener must be a function.'
      )
    }
  }

  _getOrCreateEventListeners (eventName) {
    if (!this._hasEvent(eventName)) {
      this._events.set(eventName, new Set())
    }

    return this._getListenersOfEvent(eventName)
  }

  _getListenersOfEvent (eventName) {
    return this._events.get(eventName)
  }

  _hasEvent (eventName) {
    return this._events.has(eventName)
  }
}
