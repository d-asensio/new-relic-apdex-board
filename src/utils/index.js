export * from './event-bus'
export * from './unique-id'
