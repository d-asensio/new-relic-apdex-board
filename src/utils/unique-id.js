let idCounter = -1

export function uniqueId (prefix = '') {
  return `${prefix}${++idCounter}`
}
