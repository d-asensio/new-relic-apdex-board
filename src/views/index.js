export * from './dashboard.view.js'
export * from './host.view.js'
export * from './toggle.view.js'
