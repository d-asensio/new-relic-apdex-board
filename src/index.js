
import './styles/index.scss'

import { DashboardController } from './controllers'
import { DashboardService } from './services'
import { DashboardView } from './views'

const controller = new DashboardController(
  new DashboardService(),
  new DashboardView()
)

function renderApp () {
  const appRoot = document.getElementById('app')
  controller.renderToDom(appRoot)
}

;(async function IIFE () {
  await controller.init()

  if (document.readyState === 'complete') {
    renderApp()
  } else {
    document.addEventListener(
      'DOMContentLoaded',
      () => renderApp()
    )
  }
})()
