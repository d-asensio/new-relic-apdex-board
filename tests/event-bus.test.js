import { EventBus } from '../src/utils'

describe('EventBus', () => {
  it('is created', () => {
    const eventBus = new EventBus()

    expect(eventBus).toBeInstanceOf(EventBus)
  })

  it('throws an error if the passed listener to "eventBus.on()" is not a function', () => {
    const eventBus = new EventBus()

    expect(() => {
      eventBus.on('invalidListener', null)
    }).toThrowErrorMatchingSnapshot()
  })

  it('throws an error if the passed listener to "eventBus.off()" is not a function', () => {
    const eventBus = new EventBus()

    expect(() => {
      eventBus.off('invalidListener', null)
    }).toThrowErrorMatchingSnapshot()
  })

  it('calls attached listener', async () => {
    const listener = jest.fn()
    const eventBus = new EventBus()

    eventBus.on('event', listener)
    await eventBus.emit('event')

    expect(listener.mock.calls.length).toBe(1)
  })

  it('calls attached listener with data', async () => {
    const listener = jest.fn()
    const eventBus = new EventBus()

    const dataParam1 = {}
    const dataParam2 = 'second paramenter'

    eventBus.on('event', listener)
    await eventBus.emit('event', dataParam1, dataParam2)

    expect(listener.mock.calls[0][0]).toBe(dataParam1)
    expect(listener.mock.calls[0][1]).toBe(dataParam2)
  })

  it('calls multiple attached listeners', async () => {
    const listener = jest.fn()
    const listener2 = jest.fn()
    const eventBus = new EventBus()

    eventBus.on('event', listener)
    eventBus.on('event', listener2)

    await eventBus.emit('event')

    expect(listener.mock.calls.length).toBe(1)
    expect(listener2.mock.calls.length).toBe(1)
  })

  it('do not call a attached event if unregistered trough the retuned unregister function', async () => {
    const listener = jest.fn()
    const eventBus = new EventBus()

    const unregister = eventBus.on('event', listener)
    unregister()

    await eventBus.emit('event')

    expect(listener.mock.calls.length).toBe(0)
  })

  it('do not call a attached event if unregistered trough "eventBus.off()"', async () => {
    const listener = jest.fn()
    const eventBus = new EventBus()

    eventBus.on('event', listener)
    eventBus.off('event', listener)

    await eventBus.emit('event')

    expect(listener.mock.calls.length).toBe(0)
  })

  it('do not fail when emitting unexisting event', async () => {
    const eventBus = new EventBus()

    await eventBus.emit('event')
  })

  it('do not fail when removing unexisting event', async () => {
    const eventBus = new EventBus()

    eventBus.off('event', () => {})
  })
})
