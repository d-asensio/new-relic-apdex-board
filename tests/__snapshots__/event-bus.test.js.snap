// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventBus throws an error if the passed listener to "eventBus.off()" is not a function 1`] = `"The event listener must be a function."`;

exports[`EventBus throws an error if the passed listener to "eventBus.on()" is not a function 1`] = `"The event listener must be a function."`;
